homeassistant/components/xiaomi_miio/gateway.py

```diff
@@ -0,0 +1,44 @@
+"""Code to handle a Xiaomi Gateway."""
+import logging
+
+from miio import DeviceException, gateway
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConnectXiaomiGateway:
+    """Class to async connect to a Xiaomi Gateway."""
+
+    def __init__(self):
+        """Initialize the entity."""
+        self._gateway_device = None
+        self._gateway_info = None
+
+    @property
+    def gateway_device(self):
+        """Return the class containing all connections to the gateway."""
+        return self._gateway_device
+
+    @property
+    def gateway_info(self):
+        """Return the class containing gateway info."""
+        return self._gateway_info
+
+    async def async_connect_gateway(self, host, token):
+        """Connect to the Xiaomi Gateway."""
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        try:
+            self._gateway_device = gateway.Gateway(host, token)
+            self._gateway_info = self._gateway_device.info()
```

Comment:
If this does I/O we need to schedule it on the executor thread pool with `hass.async_add_executor_job`.

Reply:
I call this function in 2 places:
- in the __init__.py file in the async_setup_gateway_entry, there I have the `hass `class so no problem
- in the config_flow.py file, there I do not have the `hass` class, how can I then call hass.async_add_executor_job?

What I did before was call the function of the gateway class with `hass.async_add_job` which schould have been `hass.async_add_executor_job`, in the __init__.py file, but I just removed that because of your remark about it above in this revieuw.

How schould I resolve this?

Reply:
In the config flow we have access to `self.hass`.

Reply:
Is it only the `info` call that does I/O or also the instantiation of `Gateway`?

Reply:
I just checked it and indeed only the info call does I/O, the instantation of Gateway only initializes the class and all of its (many) sub classes that contain all functions and info variables needed to communicate with the device, but nothing is actually sent to the device.

Handshaking, retrieving communication tokens, ids etc. is done at the first command that is sent, so in this case that is indeed at the info call that actually communicates with the device.

Reply:
Therefore I will only schedule the info call on the executor thread pool.